tools/language-server/logstat/src/main.rs

```diff
@@ -0,0 +1,317 @@
+
+    /// Number of iterations to skip
+    #[clap(long, default_value = "0")]
+    skip_iterations: usize,
+
+    /// Calculate median instead of mean
+    #[clap(long)]
+    median: bool,
+}
+
+/// Specification contains lines to lookup in the log file
+#[derive(Debug)]
+struct Spec {
+    matches: Vec<String>,
+}
+
+/// Timed operation
+#[derive(Debug)]
+struct Operation {
+    duration:  Duration,
+    timestamp: OffsetDateTime,
+    line:      String,
+}
+
+impl fmt::Display for Operation {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "{}ms [{}] {}...",
+            self.duration.whole_milliseconds(),
+            self.timestamp.format(&Rfc3339).unwrap(),
+            self.line.chars().take(80).collect::<String>()
+        )
+    }
+}
+
+/// Single iteration containing multiple sub-operations
+#[derive(Debug)]
+struct Iteration {
+    pub operations: Vec<Operation>,
+}
+
+impl Iteration {
+    pub fn total_time(&self) -> Duration {
+        let mut total: Duration = Duration::ZERO;
+        for operation in &self.operations {
+            total += operation.duration
+        }
+        total
+    }
+}
+
+/// Final statistics about benchmarked operation
+#[derive(Debug)]
+struct Stats {
+    min:  Duration,
+    max:  Duration,
+    avg:  Duration,
+    line: String,
+}
+
+impl fmt::Display for Stats {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "{avg}ms [{min}..{max}] {line}",
+            avg = self.avg.whole_milliseconds(),
+            min = self.min.whole_milliseconds(),
+            max = self.max.whole_milliseconds(),
+            line = self.line.chars().take(80).collect::<String>()
+        )
+    }
+}
+
+lazy_static! {
+    static ref RE_LOG: Regex = Regex::new(r"\[([\w]+)\] \[([\w\d:.-]+)\] (.*)").unwrap();
+    static ref WSTEST_SPEC: Spec = Spec {
+        matches: vec![
+            "wstest sent bench request".to_string(),
+            "wstest handled response".to_string(),
+        ],
+    };
+}
+
+/// Read the file with specification
+async fn read_specs(path: &PathBuf) -> Result<Spec> {
+    let file = File::open(path).await?;
+    let mut lines = BufReader::new(file).lines();
+    let mut matches = Vec::new();
+
+    while let Some(line) = lines.next_line().await? {
+        matches.push(line);
+    }
+
+    Ok(Spec { matches })
+}
+
+/// Extract the iterations information from logfile according to the provided spec
+async fn read_logfile(path: &PathBuf, spec: &Spec) -> Result<Vec<Iteration>> {
+    let file = File::open(path).await?;
+    let mut lines = BufReader::new(file).lines();
+
+    let mut iterations = vec![];
+    let mut current_operations = vec![];
+    let mut matches = spec.matches.iter();
+
+    let mut current_match = if let Some(m) = matches.next() {
+        m
+    } else {
+        matches = spec.matches.iter();
+        matches.next().expect("Empty spec!")
+    };
```

Comment:
This is very strange. We try to get first match, and when failed we try it second time?

Reply:
You're right, it can be simplified